torch/csrc/jit/codegen/cuda/consume_at.cpp

```diff
@@ -0,0 +1,444 @@
+// compute at position of producer domain. Used in computeAt pass only. No
+// checking on actual producer-consumer relationship.
+size_t getConsumerPosAlignedToProducerCA(
+    TensorView* consumer,
+    TensorView* producer) {
+  // Locate consumer's position that aligns with
+  //  the producer's new compute at axis. We need broadcast axes forwarded so we
+  //  need to replay PasC as CasP will not forward braodcast dims. For example
+  //  if we have:
+  // T2[ iS22{( 3 * 1 )} ] ca_pos( 1 ) = broadcast( T1[ iS1{3} ] ca_pos( 1 )
+  // produce_pos( 1) ) CasP will have the mapping iS1{3} -> iS2{3} and PasC will
+  // have the mapping iS22{( 3 * 1 )} <- iS1{3} We need the latter. Refer to
+  // NVFuserTest.FusionComplexBCast1_CUDA
+
+  auto c2p_map =
+      BestEffortReplay::replayPasC(
+          producer,
+          consumer,
+          -1,
+          // Compute at root domain may not be valid here, as all
+          // producers don't have to be able to map into consumer at
+          // max producer position. Since computeAt should be valid
+          // and this mechanism is only intended to lower produce
+          // position of consumer, we can simply use the pairwise map.
+          PairwiseRootDomainMap(producer, consumer))
+          .getReplay();
+
+  // Find the innermost position of consumer that has
+  //  been mapped within the producer ca axis.
+  unsigned int consumer_pos = consumer->nDims();
+  while (consumer_pos > 0) {
+    auto consumer_id = consumer->axis((int)consumer_pos - 1);
+    auto p_dom = producer->domain()->domain();
+    if (std::any_of(
+            p_dom.begin(),
+            p_dom.begin() + producer->getComputeAtPosition(),
+            [&consumer_id, &c2p_map](IterDomain* p_id) {
+              auto c_id_it = c2p_map.find(consumer_id);
+              if (c_id_it != c2p_map.end()) {
+                return c_id_it->second == p_id;
+              }
+              return false;
+            })) {
+      break;
+    }
+    consumer_pos--;
+  }
+
+  return consumer_pos;
+}
+
+void ConsumeAt::computeMaxProducerPos() {
+  std::unordered_set<TensorView*> todo;
+  for (auto p : consume) {
+    auto consumers = ir_utils::consumerTvsOf(p);
+    std::copy(
+        consumers.begin(), consumers.end(), std::inserter(todo, todo.end()));
+  }
+  for (auto tv : todo) {
+    auto producers = ir_utils::producerTvsOf(tv);
+    size_t max_pos = 0;
+    for (auto p : producers) {
+      max_pos =
+          std::max<size_t>(max_pos, getConsumerPosAlignedToProducerCA(tv, p));
+    }
+    tv->setMaxProducer(max_pos, true);
+  }
+}
+
+bool ConsumeAt::shouldPropagate(TensorView* tv) {
+  if (consume.count(tv)) {
+    unvisited.erase(tv);
+    return true;
+  }
+
+  // If one of tv's producer is in the consume set, then tv must also be
+  // replayed to obtain a compatible loop structure so that this producer
+  // can be consumed in this loop.
+  auto def = tv->definition();
```

Comment:
There's a utility function to get the producer tv's of another tv. https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/ir_utils.h#L214

Reply:
fixed